pkg/util/vsphere/storage/storage_policy.go

```diff
@@ -0,0 +1,337 @@
+
+	polv1 "github.com/vmware-tanzu/vm-operator/external/vsphere-policy/api/v1alpha1"
+)
+
+// GetStoragePolicyStatus returns the storage policy status for a given profile
+// ID.
+func GetStoragePolicyStatus(
+	ctx context.Context,
+	k8sClient ctrlclient.Client,
+	vimClient *vim25.Client,
+	pbmClient *pbm.Client,
+	profileID string) (polv1.StoragePolicyStatus, error) {
+
+	if ctx == nil {
+		panic("ctx is nil")
+	}
+	if k8sClient == nil {
+		panic("k8sClient is nil")
+	}
+	if vimClient == nil {
+		panic("vimClient is nil")
+	}
+	if pbmClient == nil {
+		panic("pbmClient is nil")
+	}
+	if profileID == "" {
+		panic("profileID is empty")
+	}
+
+	var (
+		status polv1.StoragePolicyStatus
+		pid    = []pbmtypes.PbmProfileId{
+			{
+				UniqueId: profileID,
+			},
+		}
+	)
+
+	// Get the profile and its constraints.
+	p, c, err := getCapabilityProfileAndConstraints(
+		ctx,
+		pbmClient,
+		pid,
+		profileID)
+	if err != nil {
+		return polv1.StoragePolicyStatus{}, err
+	}
+
+	// Get the information from the constraints.
+	if err := parseConstraints(c, &status); err != nil {
+		return polv1.StoragePolicyStatus{}, err
+	}
+
+	// Determine if the policy supports encryption.
+	if err := supportsEncryption(
+		ctx,
+		pbmClient,
+		profileID,
+		&status); err != nil {
+
+		return polv1.StoragePolicyStatus{}, err
+	}
+
+	// Determine which datastore types use this policy.
+	if err := getAssociatedDatastores(
+		ctx,
+		vimClient,
+		pbmClient,
+		pid,
+		profileID,
+		&status); err != nil {
+
+		return polv1.StoragePolicyStatus{}, err
+	}
+
+	// Collect unique storage class names.
+	if err := getStorageClassNames(
+		ctx,
+		k8sClient,
+		p,
+		profileID,
+		&status); err != nil {
+
+		return polv1.StoragePolicyStatus{}, err
+	}
+
+	// Sort the slices in the status.
+	status.Sort()
+
+	return status, nil
+}
+
+func getStorageClassNames(
+	ctx context.Context,
+	k8sClient ctrlclient.Client,
+	p *pbmtypes.PbmCapabilityProfile,
+	profileID string,
+	status *polv1.StoragePolicyStatus) error {
+
+	uniqueStorageClassNameMap := map[string]struct{}{
+		p.K8sCompliantName:                  {},
+		p.K8sCompliantName + "-latebinding": {},
+	}
+	for _, n := range p.OtherK8sCompliantNames {
+		uniqueStorageClassNameMap[n] = struct{}{}
+	}
+	for n := range uniqueStorageClassNameMap {
+		var (
+			obj storagev1.StorageClass
+			key = ctrlclient.ObjectKey{Name: n}
+		)
+		if err := k8sClient.Get(ctx, key, &obj); err != nil {
+			if err := ctrlclient.IgnoreNotFound(err); err != nil {
+				return fmt.Errorf(
+					"failed to get storage class %q for profile %q: %w",
+					n, profileID, err)
+			}
+		} else {
+			status.StorageClasses = append(status.StorageClasses, n)
+		}
+	}
+	return nil
+}
+
+func getAssociatedDatastores(
+	ctx context.Context,
+	vimClient *vim25.Client,
+	pbmClient *pbm.Client,
+	pid []pbmtypes.PbmProfileId,
+	profileID string,
+	status *polv1.StoragePolicyStatus) error {
+
+	assocEnts, err := pbmClient.QueryAssociatedEntities(ctx, pid)
```

Comment:
I don't think this does what you intended to do. This most likely returns the associated vms/disks/datastores for the profile. When it comes to datastores it returns the associated default requirement profile, I don't think that's what you want, perhaps you intend to return compatible datastore? i think the API for that is checkRequirements

Reply:
You are correct sir, thank you. Let me update this. Wow, such a great catch!!